clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java

```diff
@@ -1531,6 +1534,9 @@ private Fetch<K, V> pollForFetches(Timer timer) {
             return fetch;
         }
 
+        // send any new fetches (won't resend pending fetches)
+        sendFetches(timer, true);
```

Comment:
Similar concern as in the other call to sendFetches. I wonder if this 'true' here could be dangerous in cases of poll(0/lowTimeout)? In those cases, we could easily get a `TimeoutException` in this line because of the `addAndGet`, and we would break out of here without making it to awaitNotEmpty right? I guess that the overall situation would be sorted out on the next poll iteration, that would find the fetch results available, but still doesn't feel right to break on Timeout on the "generate request" phase, instead of on the "awaitNotEmpty" phase.

At this point I'm only spotting a difference that doesn't feel conceptually right, but honestly not sure if the implications are truly harmful in this case. The alternative I see is to pass shouldWait here false to ensure we don't block when telling the background to generate fetches, then the awaitNotEmpty is the one to apply the timer. What do you think?

Reply:
True. `ClassicKafkaConsumer.sendFetches()` does not make use of a `Timer` or timeout of any sort, so the act of creating the outgoing requests shouldn't yield any timeouts.

`AsyncKafkaConsumer.sendFetches()` does take a `Timer`, only because we do want to wait for the requests to be created. And the easiest way to wait for the requests is to use the `addAndGet()` approach. If we were to switch to `add()`, we wouldn't _technically_ be waiting for the requests to be created, but I'm not sure that's really that much of a concern in practice.

So in the case you mentioned (small timeout values), if we send the `CreateFetchRequestsEvent` but don't wait for it, then we shift the possibility of the `TimeoutException` to the `FetchBuffer.awaitNotEmpty()` call. I don't think that's a problem.

So, correct me if I'm wrong, but it sounds like we want to change `CreateFetchRequestsEvent` so that it isn't a `CompletableApplicationEvent` since we don't want to wait in either case of fetch requests: blocking or prefetch. I'm fine with that and it'll make the code a little cleaner too.

Reply:
So I realized I couldn't (comfortably) remove the wait because one of the byproducts is the ability to catch errors and bubble them up. PTAL at the latest (with probably too many comments) to see what you think. Thanks!